### ISSUE TITLE
[promise-function-async] <should ignore inline functions passed to then / catch>

### ISSUE THREAD

**Issue by kasvtv** [typescript-eslint/typescript-eslint#1708]
I love to be reminded to mark any function that returns a promise async. The reason being, if it throws an error it will of course return a rejected promise instead of actually throw.

However, when passing an inline function to a `.then` or `.catch` handler, marking it async is not necessary since it will be handled identically whether the function actually throws or returns a promise.

So I don't want to be reminded to mark it async in that case. Maybe this should be an option?

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/promise-function-async": ["error"]
  }
}
```

```TS
const a = async () => 2
const b = async () => Promise.resolve().then(() => a())
```

**Expected Result**
No issues reported
**Actual Result**
Functions that return promises should be async
**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.22.0` |
| `@typescript-eslint/parser`        | `2.22.0` |
| `TypeScript`                       | `3.8.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `12.13.0` |
| `npm`                              | `6.12.0` |

**Comment by bradzacher** [comment]
IMO it should be handled the same way regardless.

- It's about consistency across the codebase.
- I don't think it's right to add a special case just because in this specific context, the function might do the same thing based on an implementation detail.
- This sort of branch of logic makes the rule work in a non-intuitive way, which can be confusing for end users. For example, if you refactor the arrow function outside of the `then`, then it starts to error. As a user, it would not be easy to understand why that is without specifically reading the docs, which makes the rule harder to trust.
- Additionally, it's a non-trival branch of logic to check for this specific case in this specific context.

For the above reasons, I'm going to reject this proposal.

---

As a sidebar - why are you using async, but not await?
```ts
const a = async () => 2
const b = async () => {
  await Promise.resolve()
  return a();
}
```

**Comment by kasvtv** [comment]
Perhaps my example was too contrived.
Usage of async in the `b` previous example was not to have this same rule pop up and I was not using await to demonstrate having a `.then()` clause.

Let me rephrase:
I like this rule because requiring functions marked async makes code more predictable and easier to navigate. However, any functions inside then and catch bindings clauses attached to a promise are pretty much by definition going to deal with asynchronicity, making it seem superfluous to me to add `async` there. From what I've seen from the community, marking something `async` that doesn't necessarily use `await`, but just returns a promise, is quite common. Having `async` in your then and catch bindings is quite unusual and goes against the way most people are used to read and write code.

**Comment by bradzacher** [comment]
I understand what you're saying, but I think it's weirder to have this contextually unnecessary async keyword.

For example, if you create a function that accepts a `Promise<string> | string`, should this function also not require the `async` keyword, because it handles both promises and non-promises equally? The answer is probably no, right?

And this leads into the 3rd point in my above comment - it's not intuitive why a function that's passed to then/catch doesn't require an async - it's a special case for the sake of a special case.
It's not easy for a user to intuit why a function triggers a lint rule outside the then, but not when it's inside it.
And if you test it against a similar function, then a user would just end up with "oh it's just how the rule works", which damages trust in the linting tooling.

> goes against the way most people are used to read and write code

This is something that changes pretty quickly with lint rules though, right?
It only takes a few hits of the rule to update your mental model of how to write in the codebase, and then because it's done everywhere, it becomes the way you read it.

Some people don't like this rule for the same reason - they think it's not  correct to write an async function that doesn't use await. But most people like it because it creates the consistent marker for "this function always returns a promise".
